Builtin function to parse uuid with google/uuid library

Od1nB opened the pull request:
Signed-off-by: Odin Bjørnebo <odinbjoernebo@gmail.com>

<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see our contributor guide below.

For more information on contributing to OPA see:

* [Contributing Guide](https://www.openpolicyagent.org/docs/latest/contributing/)
  for high-level contributing guidelines and development setup.

-->

### Why the changes in this PR are needed?

<!--
Include a short description of WHY the changes were made.
-->

### What are the changes in this PR?

<!--
Include a short description of WHAT changes were made.
-->

### Notes to assist PR review:

<!--
Here you can add information you think will help the reviewer(s).
-->

### Further comments:

<!--
Here you can include links to additional resources related to the changes, discuss your solution, other approaches you considered etc.
-->

yogisinha committed:
Changed the LoadPaths function.
Squashing all the commits for following issue:
Fix for the issue when OPA doesnot load tarball on cmd line as a bundle.

Fixes #5879

Signed-off-by: Yogesh Sinha <sinhayogi@gmail.com>

gitu committed:
remove not required basedir for oci bundles & add test to verify signature verification

Signed-off-by: Florian Schrag <f@schr.ag>

c2zwdjnlcg committed:
Support force cache even when server doesn't set the Date header

This fixes a bug where when a upstream server doesn't set a Date
header the response will be cached but then once the cache is
expired it never gets refreshed because parseResponseHeaders
returns a no date header error

This also makes it so that the force_cache_duration_seconds
is respected when an upstream returns a 304 not modified

Signed-off-by: Peter <c2zwdjnlcg@users.noreply.github.com>

johndbro1 committed:
doc: updated support page to include a link to the OPA support page at PACLabs

Signed-off-by: John Brothers <johnbr@gmail.com>

johanfylling committed:
ast: Making partial object key rules contribute to dynamic portion of object type (#6177)

... instead of static portion.

This change allows object unification/comparison between an object composed at eval-time (i.e. constructed by rules) and a static object that doesn't contain all keys declared by rules, as those might not be defined at eval-time.

Fixes: #6138
Signed-off-by: Johan Fylling <johan.dev@fylling.se>

slonka committed:
feat(download): add opa_no_oci flag to build without containerd

Signed-off-by: slonka <slonka@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump github.com/containerd/containerd from 1.7.3 to 1.7.4

Bumps [github.com/containerd/containerd](https://github.com/containerd/containerd) from 1.7.3 to 1.7.4.
- [Release notes](https://github.com/containerd/containerd/releases)
- [Changelog](https://github.com/containerd/containerd/blob/main/RELEASES.md)
- [Commits](https://github.com/containerd/containerd/compare/v1.7.3...v1.7.4)

---
updated-dependencies:
- dependency-name: github.com/containerd/containerd
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

anderseknert committed:
docs: make it clear dropped decisions aren't logged (#6180)

And modernize the Rego examples while at it.

Signed-off-by: Anders Eknert <anders@styra.com>

joaobrandt committed:
cmd: Fix opa test --ignore when used together with --bundle

When using the --bundle flag, the opa test was ignoring (pun intented)
the --ignore flags.

Signed-off-by: João Paulo G. Brandt <joao@miro.com>

srenatus committed:
rego: expose PrepareOption, add BuiltinFuncs

This is used as a mechanism to inform custom rego target plugins about builtins
that have been defined ad-hoc, via rego.Function1, rego.Function2, etc.

Before, in a setup like this:

```go
	regoArgs = append(regoArgs,
		rego.Target("my_custom_evaluator"),
		// ...
		rego.Function2(builtins.RegalParseModuleMeta, builtins.RegalParseModule),
		rego.Function1(builtins.RegalJSONPrettyMeta, builtins.RegalJSONPretty),
		rego.Function1(builtins.RegalLastMeta, builtins.RegalLast),
	)
```
...the prepared eval state for the custom rego target would not be informed about the
functions defined there.

Now, the PrepareOptions that are passed to the target plugins are used to carry the
information along. To do that, a new field was added, and an accessor function.

In the working stages of this, I had also added accessors for NoInline and PartialEval,
but since the target plugin mechanism explicitly does not support anything PE-related
(yet), I've rolled those changes back.

Signed-off-by: Stephan Renatus <stephan@styra.com>

johanfylling committed:
ast: Including "child" rules when fetching rules by ref (#6183)

* ast: Including "child" rules when fetching rules by ref

In order to properly sort rules according to dependencies before performing type-checking.

Also, properly merging types when type-tree produces multiple types for a given ref.

Fixes: #6182
Signed-off-by: Johan Fylling <johan.dev@fylling.se>

srenatus committed:
golang: 1.20.7 -> 1.21 (#6189)

https://tip.golang.org/doc/go1.21

This required some test updates:

* topdown/tokens_test: adjust for go1.21

   What was correct for 1.20 is correct for 1.21, so I've flipped the exception logic.

* plugins/rest/TestClientCert: adapt cert-related error string

* internal/prometheus/TestJSONSerialization: add new metrics

   There are new metrics!

Signed-off-by: Stephan Renatus <stephan@styra.com>

sspaink committed:
topdown: add numbers.range_step built-in function (#6187)

Fixes: #6186
Signed-off-by: Sebastian Spaink <sebastianspaink@gmail.com>

mrueg committed:
chore: Replace ghodss/yaml with sigs.k8s.io/yaml (#6195)

This is a fork that is maintained by the kubernetes community.

https://github.com/kubernetes-sigs/yaml

Signed-off-by: Manuel Rüger <manuel@rueg.eu>

dependabot[bot] committed:
build(deps): bump oras.land/oras-go/v2 from 2.2.1 to 2.3.0

Bumps [oras.land/oras-go/v2](https://github.com/oras-project/oras-go) from 2.2.1 to 2.3.0.
- [Release notes](https://github.com/oras-project/oras-go/releases)
- [Commits](https://github.com/oras-project/oras-go/compare/v2.2.1...v2.3.0)

---
updated-dependencies:
- dependency-name: oras.land/oras-go/v2
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
[docs] Move OPA ecosystem pages to site top-level (#6198)

* We are having trouble managing the complexity of having ecosystem pages shown in the versioned area of the OPA docs site (see website: link ecosystem from edge #6170)
* We have invested in the OPA ecosystem data recently and it makes sense to make it more prominent.
* The data in the OPA Ecosystem is harder to use when nested in docs as some Hugo functions are unavailable to nested sections.

This PR implements a solution by:

* Pages that were under /docs/version/ecosystem, /docs/version/integrations etc have been moved to /ecosystem and /integrations.
* Redirects have been created for the old pages in the netlify config.

Signed-off-by: Charlie Egan <charlie@styra.com>

johanfylling committed:
General refs in rule heads (#5913)

* Adding support for multiple variables at arbitrary locations in rule refs
* Updating type-checker to handle general ref heads

Fixes: #5993
Fixes: #5994

Signed-off-by: Johan Fylling <johan.dev@fylling.se>

Commit:
[docs] Add intro text to Ecosystem sections (#6201)


Signed-off-by: Charlie Egan <charlie@styra.com>

Commit:
[docs] Address issues with Ecosystem links in navs (#6200)

Signed-off-by: Charlie Egan <charlie@styra.com>

johanfylling committed:
Prepare v0.56.0 release (#6202)

Signed-off-by: Johan Fylling <johan.dev@fylling.se>

johanfylling committed:
Prepare v0.57.0 development

Signed-off-by: Johan Fylling <johan.dev@fylling.se>

Od1nB committed:
Builtin function to parse uuid with google/uuid library

Fixes #6173

Signed-off-by: Odin Bjørnebo <odinbjoernebo@gmail.com>

kristiansvalland commented on topdown/uuid.go:
```diff
@@ -31,6 +32,32 @@ func builtinUUIDRFC4122(bctx BuiltinContext, operands []*ast.Term, iter func(*as
+
+	result := ast.NewObject()
+	for prop, val := range parsed {
+		switch parsed[prop].(type) {
```
Use `switch val := val.(type) {` so that you do not have to cast one more time in the body.

Od1nB replied:
oooh, that'll look better yea

Od1nB replied:
fixed with amend

kristiansvalland commented on topdown/uuid.go:
```diff
@@ -31,6 +32,32 @@ func builtinUUIDRFC4122(bctx BuiltinContext, operands []*ast.Term, iter func(*as
+		case int:
+			result.Insert(ast.StringTerm(prop), ast.IntNumberTerm(val.(int)))
+		case int64:
+			result.Insert(ast.StringTerm(prop), ast.IntNumberTerm(int(val.(int64))))
```
Potential overflow? Or what year will unix nano overflow assuming int32 architecture?

Od1nB replied:
[wiki](https://en.wikipedia.org/wiki/Year_2038_problem), that would be year 2038, would that still be a problem since it is defined as a int64?

kristiansvalland replied:
I think the year 2038 is for unix stored as seconds using a signed 32-bit integer. However, I think that unixNano will already overflow with todays date. At time of writing, the unix time is 1693809950, which in ns is `1 693 809 950 000 000 000`, which is `0b1011110000001100111110010110011100010110001011010110000000000`, or 61 bits. So int, assuming it uses 32 bits, will definitely overflow.

Instead, something like this can be done: https://github.com/open-policy-agent/opa/blob/main/topdown/time.go#L89
```go
ast.NumberTerm(int64ToJSONNumber(int64(value)))
```

Od1nB replied:
You are completely correct. Didn't look at the sheer size of ns. That is way over 32bits 😮 Yep seems like the `int64toJSONNumber(val int64)` will work. I'll change the insert

kristiansvalland commented on topdown/uuid.go:
```diff
@@ -31,6 +32,32 @@ func builtinUUIDRFC4122(bctx BuiltinContext, operands []*ast.Term, iter func(*as
+		return err
+	}
+
+	parsed := uuid.Parse(string(str))
```
I am wondering if it would have been better that uuid.Parse returns `(map[string]any, error)` instead of just `map[string]any`. It would make it more obvious that it is on parse errors that we return `undefined`

Od1nB replied:
```
parsed := uuid.Parse(string(str))
if len(parsed) == 0 {
     return nil
}
```
riight, so it would be more explicit to return an error, check for that and then return the undefined? We dont want the parsing to cause an error in the application, so it would just check for an error and not return it. The code above is how it is handled in the topdown function now

kristiansvalland commented on test/cases/testdata/uuid/test-uuid-parse.yaml:
```diff
@@ -0,0 +1,27 @@
+cases:
```
Potential test cases that should be added are for 
- URN-UUIDs: `urn:uuid:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`
- UUIDs in the format `{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}`
- UUIDs in the format `xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx`

Just to verify that the builtin is in parity with the google/uuid library.

Ref: https://github.com/google/uuid/blob/b3cae7c3064a8404a9f1eb0bb0662787ca3144f3/uuid.go#L64

Od1nB replied:
smart smart, I´ll make some cases for it

Od1nB replied:
made tests, updated with amend

kristiansvalland commented on internal/uuid/uuid.go:
```diff
@@ -20,3 +26,67 @@ func New(r io.Reader) (string, error) {
+// with some decoded values with fillMap
+func Parse(s string) map[string]interface{} {
+	uuid, err := uuid.Parse(s)
+	out := make(map[string]interface{})
```
minuscule performance improvement: The size of the map can be precomputed.
- 1: 5
- 2: 7
- 3, 4, 5: 2

Od1nB replied:
made case for 1 and 2, the rest could be default. If version 6 comes out it will at least have length 2 and version and variant

kristiansvalland commented on internal/uuid/uuid.go:
```diff
@@ -20,3 +26,67 @@ func New(r io.Reader) (string, error) {
+// Fills the map with values from the uuid. Version and variant for every version.
+// Version 1-2 has decodable values that could be of use, version 4 is random,
+// and version 3,5 is not feasible to extract data.
+func fillMap(m map[string]interface{}, u *uuid.UUID) {
```
nit: I do not think `u` needs to be a pointer, since the underlying type is `[16]byte` which is already a pointer type.

kristiansvalland commented on capabilities.json:
```diff
@@ -4559,6 +4559,28 @@
+        "result": {
+          "dynamic": {
+            "key": {
+              "type": "any"
```
Is this the key in the result map? If so, can't this be "string"?

Od1nB replied:
Yes, it should be a string

kristiansvalland commented on builtin_metadata.json:
```diff
@@ -18256,6 +18257,26 @@
+    ],
+    "description": "Parses incoming string if valid uuid, returns object with properties if valid",
+    "introduced": "edge",
+    "result": {
+      "description": "another out",
+      "name": "out",
+      "type": "object[any: any]"
+    },
```
I may misinterpret how this will end up looking on the web page, but it seems like the "description" and "name" can mo improved a little. Also, the type is `object[string: any]`

kristiansvalland commented on builtin_metadata.json:
```diff
@@ -18256,6 +18257,26 @@
+    "available": [
+      "edge"
+    ],
+    "description": "Parses incoming string if valid uuid, returns object with properties if valid",
```
Suggestion: "Parses the string value as an UUID and returns an object with the well-defined fields of the UUID if valid".
This suggestion also applies to the `description` in `var UUIDParse`

Od1nB replied:
All these are fair and, should be changed into something that would look better on the webpage. And also explained the different output variables. Think these were generated with `make generate`, so thats why the desc in `var UUIDParse` is similiar

Od1nB replied:
changed it into your suggestion and made it look more like other builtin docs

kristiansvalland reviewed (commented):
Really nice feature addition! I do not have any concerns about the implementation, just some thoughts and optional suggestions :)

dependabot[bot] committed:
build(deps): bump aquasecurity/trivy-action from 0.11.2 to 0.12.0 (#6207)

Bumps [aquasecurity/trivy-action](https://github.com/aquasecurity/trivy-action) from 0.11.2 to 0.12.0.
- [Release notes](https://github.com/aquasecurity/trivy-action/releases)
- [Commits](https://github.com/aquasecurity/trivy-action/compare/0.11.2...0.12.0)

---
updated-dependencies:
- dependency-name: aquasecurity/trivy-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
